tools/ChangelogTools.psm1

```diff
@@ -0,0 +1,401 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+#requires -Version 6.0
+
+using module .\GitHubTools.psm1
+
+class IgnoreConfiguration
+{
+    [string[]]$User
+    [string[]]$IssueLabel
+    [string[]]$PRLabel
+    [string[]]$CommitLabel
+}
+
+class ChangeInfo
+{
+    [GitHubCommitInfo]$Commit
+    [GitHubPR]$PR
+    [GitHubIssue[]]$ClosedIssues
+    [int]$IssueNumber = -1
+    [int]$PRNumber = -1
+    [string]$ContributingUser
+    [string]$BodyText
+    [string]$Subject
+}
+
+class ChangelogEntry
+{
+    [uri]$IssueLink
+    [uri]$PRLink
+    [string]$Category
+    [string[]]$Tags
+    [string]$BodyText
+    [string]$Subject
+    [string]$Thanks
+    [string]$RepositoryName
+    [ChangeInfo]$Change
+}
+
+class ChangeLog
+{
+    ChangeLog()
+    {
+        $this.Sections = [System.Collections.Generic.Dictionary[string, ChangelogEntry]]::new()
+    }
+
+    [string]$ReleaseName
+    [datetime]$Date
+    [string]$Preamble
+    [System.Collections.Generic.Dictionary[string, ChangelogEntry]]$Sections
+}
+
+function NormalizeSubject
+{
+    [OutputType([string])]
+    param(
+        [Parameter(Mandatory)]
+        [string]
+        $Subject
+    )
+
+    $Subject = $Subject.Trim()
+    if ([char]::IsLower($Subject[0])) { $Subject = [char]::ToUpper($Subject[0]) + $Subject.Substring(1) }
+    if ($Subject[$Subject.Length] -ne '.') { $Subject += '.' }
+
+    return $Subject
+}
+
+filter Get-ChangeInfoFromCommit
+{
+    [OutputType([ChangeInfo])]
+    param(
+        [Parameter(Mandatory, ValueFromPipeline, Position=0)]
+        [GitHubCommitInfo[]]
+        $Commit,
+
+        [Parameter(Mandatory)]
+        [string]
+        $GitHubToken
+    )
+
+    foreach ($singleCommit in $Commit)
+    {
+        Write-Verbose "Getting change information for commit $($Commit.Hash)"
+
+        $changelogItem = [ChangeInfo]@{
+            Commit = $singleCommit
+            BodyText = $singleCommit.Body
+            Subject = $singleCommit.Subject
+            ContributingUser = $singleCommit.GitHubCommitData.author.login
+        }
+
+        if ($Commit.PRNumber -ge 0)
+        {
+            $getPrParams = @{
+                Organization = $singleCommit.Organization
+                Repository = $singleCommit.Repository
+                PullNumber = $singleCommit.PRNumber
+                GitHubToken = $GitHubToken
+            }
+            $pr = Get-GitHubPR @getPrParams
+
+            $changelogItem.PR = $pr
+            $changelogItem.PRNumber = $pr.Number
+
+            $closedIssueInfos = $pr.GetClosedIssueInfos()
+            if ($closedIssueInfos)
+            {
+                $changelogItem.ClosedIssues = $closedIssueInfos | Get-GitHubIssue
+                $changelogItem.IssueNumber = $closedIssueInfos[0].Number
+            }
+        }
+
+        $changelogItem
+    }
+}
+
+filter New-ChangelogEntry
```

Comment:
I don't know how important making Codacy green is here, but could add something like: `[Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSUseShouldProcessForStateChangingFunctions", "", Justification="We're not actually changing system state.")]` to suppress it complaining about failing the PSSA rule...

Reply:
Ironically there are much more dangerous methods in the GitHubTools.psm1 module that it misses. PSScriptAnalyzer has some weird bits in it frankly -- I'd love to make it more logical.